articles/200_coexistence_and_migration.md

```diff
@@ -0,0 +1,249 @@
+If that is the case a backward compatible parser to read existing `.msg` / `.srv` files will likely be available.
+
+### Generated message / service API
+
+In order to satisfy the goal of dual-homed applications the generated code of ROS 1 and ROS 2 need to be side-by-side usable.
+
+For C++ that is already the case.
+In ROS 1 the header files are generated in namespace specific to the package name.
+In ROS 2 the header files are placed into a sub-namespace names `msg` / `srv`.
+Beside the different path / namespace the data structure also uses different types for e.g. the shared pointers (Boost vs. C++11).
+
+For Python that is not the case.
+
+**Proposal:** Change the location of the generated Python modules from using `msg` / `srv` as the subdirectory to `message` / `service`.
+At the same time updating the subdirectory name of the C++ headers would make sense to be similar.
+
+For backward compatibility the same API as for ROS 1 could be generated into a separate location.
+The user could opt-in to use these to reduce the migration effort but by sacrificing the dual-home support.
+In order to use those ROS 1 shim APIs the user would have to add another path to the include directories / `PYTHONPATH`.
+
+**Idea:** Implement these additional message generators and provide a mechanism to easily use those interfaces.
+
+## Different categories of changes
+
+For the migration two different categories of changes need to be considered separately:
+
+* Changes which can be applied to ROS 1 packages to either make it compatible with ROS 2 or minimize the necessary changes.
+* Changes which are necessary to make the package work for ROS 2 but which are not compatible with ROS 1.
+
+If the second set of changes would be empty a package could be used as a ROS 1 package as well as ROS 2 packages using the same sources.
+Currently the previous sections mention a few different items which fall into the second category therefore it doesn't seem realistic to achieve compatibility using the exact same sources.
+The following sections assume that it is not feasible to maintain a single source for the ROS 1 as well as the ROS 2 version of a package.
+
+### Changes applicable to ROS 1 packages
+
+* Update the package manifest to format 2.
+
+* In C++ introduce typedefs for all message / service classes as well as their pointer types.
+  This will allow to easily change the typedef to a different type in ROS 2.
+
+* Update Python code to also target Python 3.5+.
+
+### Changes necessary for migrating a package to ROS 2
+
+* Update the dependencies in the manifest.
+
+* Update the CMake code to find the updated dependencies.
+
+* The API of the message / service classes has changed:
+
+  * Different headers / modules to include / import.
+  * Different classes / methods / types.
+  * C++11 interface using e.g. `std::shared_ptr` rather then relying on `Boost`.
```

Comment:
I would like to see a little note about Windows compatibility, i.e. exporting headers and/or compiling into shared libraries.

Reply:
What would you like to see? I don't think this has anything to do with coexistence. Or at least boost vs. std doesn't make a difference here.

Reply:
It's not related to coexistence, but certainly with migration and maintenance. It's not related to boost or std, I wanted to add another bullet point. Because if you want to migrate your existing ROS1 code to ROS2 and want to have it run on Windows, you'll have to export the symbols.

Reply:
I would argue that this is not related to ROS 2. If you want your ROS 1 package to compile on Windows you already have to do the same. And if you migrate to ROS 2 but don't care about Windows support that is not an issue either. Therefore I would prefer not to mention it in this article.